lib/cppcheck.h

```diff
@@ -146,7 +146,7 @@ class CPPCHECKLIB CppCheck : ErrorLogger {
     bool isUnusedFunctionCheckEnabled() const;
 
     /** Remove *.ctu-info files */
-    void removeCtuInfoFiles(const std::map<std::string, std::size_t>& files);
+    void removeCtuInfoFiles(const std::map<std::string, std::size_t>& files); // cppcheck-suppress functionConst // has side effects
```

Comment:
hmm.. after your changes we now get a warning here right? Can we avoid this? In the test cases you wrote the new warnings make sense.

Reply:
The warning is correct in the sense that the code will compile after adding `const`. We can't know if the user wants to do that though, and we say so in the extended error message. But this is a general issue of this check and not related to this PR. Personally, I think we should stick with the C++ definition of `const`.

Reply:
I suggest that we will split it up. A non-inconclusive id where it does not warn if there are potential side effects. Another id and inconclusive if there are potential side effects.

The test cases you added can be non-inconclusive. It can assume that const functions do not have side effects.

Reply:
The function does filesystem changes without any file descriptors involved. Try to treat these non-`const` is borderline impossible since you could call any API in there which might modify something. Determining the access of class owned pointers will already be very difficult. I think if you really want/need that kind of safe you probably need to use object  wrappers around the given functionality in your design (like the "safe_ptr" suggested in another comment). I also did not see any function attributes which would allow for this to be tracked.

Reply:
A valid case which comes to mind where this would never be applied would be asynchronous callbacks. Maybe the method not be `const` if the given callback might modify something? It will be impossible to determine what passes for a "side effect" and even if you try to do that it might require some serious tuning flags so it might actually be useful at all in code bases which are just slightly more complex than our own.

That really seems like wasting engineering time and this check is also just "inconclusive" (and it probably has to stay that way always) and people should not treat these warnings as an end to all.

Reply:
The idea is probably to have a new warning `functionConstForReal` in case we are sure that there are a no side effects, and to emit `functionConst` otherwise. But this is beyond the scope of this PR.

Reply:
Since we are introducing lots of suppressions here and cannot really determine the "side effects" which are brought up here it kind of is in the scope.

I would not consider these side effects at all since the code is valid and that's what the spec allows and it is an inconclusive check.

I think we should make the code "safe by design" (see `safe_ptr`) as "safe by spec" is not "safe" at all.

On a side note it appears that raw pointer accesses are currently being outlawed (see clang warning `-Wunsafe-buffer-usage` along with several clang-tidy checks) so the cases this should affect should lessen. But unfortunately the smart pointers give access to non-cost raw pointers since the container is bound to the underlying type and `const` does not affect that.

Reply:
> The idea is probably to have a new warning functionConstForReal in case we are sure that there are a no side effects, and to emit functionConst otherwise. But this is beyond the scope of this PR.

Yes.

When I google I can see that some people think that if there are side effects then it's not proper to make the method const.

I would say the accepted answer here says so:
https://stackoverflow.com/questions/9807876/const-member-function-when-operation-has-clear-side-effects

And it is said in the C++FAQ that const should not be based on the physical state but the logical state of the object:
https://isocpp.org/wiki/faq/const-correctness#const-member-fns
"The trailing const on inspect() member function should be used to mean the method won’t change the object’s abstract (client-visible) state. That is slightly different from saying the method won’t change the “raw bits” of the object’s struct. "

And I also think this link is interesting:
https://isocpp.org/wiki/faq/const-correctness#logical-vs-physical-const
This checker currently focus on changes to the "physical state" while C++FAQ recommends that the "logical state" determines if the method is const or not.